File: backend/code_java/readwordexample/readwordcode/src/main/java/com/code/countword/Implementation/ReadFileImp.java
package com.code.countword.Implementation;

import com.code.countword.util.InfoWordCount;
import com.code.factory.SubscriberFactory;
import com.code.factory.SubscriberFactoryImp;

import java.io.*;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ForkJoinTask;
import java.util.concurrent.RecursiveTask;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ReadFileImp extends RecursiveTask<Long> {
    @Serial
    private static final long serialVersionUID = 1L;
    private final SubscriberFactory factory = SubscriberFactoryImp.getInstance();
    private final File files;
    private static final long NUMBER_LINES=100;
    private static final int DEFAULT_TASK=1;
    private static final int TASK=1;
    private final InfoWordCount infoWordCount;

    public ReadFileImp(File file, InfoWordCount infoWordCount) {
        this.files=file;
        this.infoWordCount = infoWordCount;
    }
    private BufferedReader getContentFile() throws FileNotFoundException {
        return new BufferedReader(new FileReader(files));
    }

    private long forkFile(long countLines, List<String> lines){
        int fromIndex=0;
        int toIndex;
        int toIndexAux;
        int numTasks= (int)Math.ceil((double)countLines/ReadFileImp.NUMBER_LINES);
        toIndex=(int)countLines/numTasks;
        toIndexAux=toIndex;
        List<RecursiveTask<Long>> fork = new LinkedList<>();
        for(int iTask = 0;iTask< numTasks;iTask++) {
            CountWordImp counter = factory.getCountWord(lines.subList(fromIndex, toIndex),files,iTask,numTasks,infoWordCount);
            fork.add(counter);
            counter.fork();
            fromIndex=toIndex;
            toIndex=toIndex+toIndexAux;
            if(iTask==(numTasks-2))toIndex=(int)countLines;
        }
        return fork.stream().map((Function<RecursiveTask<Long>, Object>) ForkJoinTask::join).count();
    }

    private Long readAllFile() throws IOException {
        long size = this.files.length();
        BufferedReader read = getContentFile();
        List<String> lines  = read.lines().collect(Collectors.toList());
        read.close();
        long countLines = lines.size();
        if(countLines!=0) {
            if(countLines>100 || size > 500000) {
                return forkFile(countLines, lines);
            }else {
                CountWordImp counter = factory.getCountWord(lines,files,DEFAULT_TASK,TASK,infoWordCount);
                counter.fork();
                return counter.join();
            }
        }
        else {
            return 0L;
        }
    }

    @Override
    protected Long compute() {
        try {
            return readAllFile();
        } catch (IOException e) {
            e.printStackTrace();
            return 0L;
        }
    }
}

File: backend/code_scala/readwordexample/readwordcode/src/main/scala/com/code/countword/implementation/ReadFileImp.scala
package com.code.countword.implementation

import com.code.countword.contract.ReadFile
import com.code.countword.util.{AbstractExecutionContext, InfoWordCount}
import com.code.factory.SubscriberFactory

import java.io.{BufferedReader, FileNotFoundException, FileReader}
import java.io.File
import scala.annotation.tailrec
import scala.concurrent.Future
import scala.jdk.CollectionConverters.CollectionHasAsScala
import scala.language.postfixOps

class ReadFileImp(file: File, infoWordCount: InfoWordCount) extends AbstractExecutionContext with ReadFile{
  private val factory = SubscriberFactory

  private val NUMBER_LINES = 100
  private val DEFAULT_TASK = 1
  private val TASK = 1
  override def readFile(): Future[Option[Long]] = Future{
      readAllFile()
    }.flatten

  @throws[FileNotFoundException]
  private def getContentFile = new BufferedReader(new FileReader(file))

  private def readAllFile():Future[Option[Long]]={
    val sizeFile = file.length
    val read = getContentFile
    val lines  = read.lines.toList
    read.close()
    val countLines = lines.size
    countLines match {
      case qtaLines if qtaLines>0 && (countLines>100 || sizeFile > 500000)=>
        forkFile(countLines,lines.asScala toList)
      case qtaLines if qtaLines>0 =>factory.getCountWord(DEFAULT_TASK,TASK, infoWordCount)
        .countWordFork(lines.asScala toList,file)
      case _ =>Future.successful(Option(0))
    }
  }

  private def forkFile(countLines: Int, lines: List[String]):Future[Option[Long]]={
    val numTasks:Int = Math.ceil(countLines/NUMBER_LINES) toInt
    val toIndex = countLines/numTasks
    @tailrec
    def _callCountWord(toIndexAux:Int, toIndex: Int=0, task: Int=0, fromIndex:Int=0):Future[Option[Long]]= task match {
      case iTask if iTask<numTasks && iTask==(numTasks-2)=>
        callCountWord(lines,fromIndex,toIndex,iTask,numTasks)
        _callCountWord(toIndex,countLines,task+1,toIndex)
      case iTask if iTask<numTasks=>
        callCountWord(lines,fromIndex,toIndex,iTask,numTasks)
        _callCountWord(toIndexAux,toIndexAux+toIndex,task+1,toIndex)
      case _ => Future.successful(Option(1))
    }
    _callCountWord(toIndex ,toIndex)
  }

  private def callCountWord(lines: List[String], fromIndex:Int, toIndex:Int, iTask:Int, numTasks:Int): Future[Option[Long]] =
    factory.getCountWord(iTask,numTasks, infoWordCount).countWordFork(lines.slice(fromIndex, toIndex),file)
}
